feat(router-core): validate params while matching (#5936)
tannerlinsley/react-location: docs/router/framework/react/guide/path-params.md

Optional path parameters provide a powerful and flexible foundation for implementing internationalization in your TanStack Router applications. Whether you prefer prefix-based or combined approaches, you can create clean, SEO-friendly URLs while maintaining excellent developer experience and type safety.

## Validating Path Parameters During Matching

> [!WARNING]
> The `skipRouteOnParseError` option is currently **experimental** and may change in future releases.

> [!IMPORTANT]
> **Performance cost**: This feature has a **non-negligible performance cost** and should not be used indiscriminately. It creates additional branches in the route matching tree, reducing matching efficiency and requiring more route evaluations. Use it only when you genuinely need type-specific routes at the same path level.

> By default, TanStack Router matches routes based purely on URL structure. A route with `/$param` will match any value for that parameter, and validation via `params.parse` happens later in the route lifecycle. If validation fails, the route shows an error state.

Reviewer: ```suggestion
By default, TanStack Router doing route matching purely base on URL structure. A route with `/$param` will match any value for that parameter, with validation via `params.parse` happening later in the route lifecycle. If validation fails, the route is set into an 'error state'.
```
